File: secret-handshake/SecretHandshake.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class SecretHandshake
{
    private static readonly IDictionary<int, string> commandValues;

    static SecretHandshake()
    {
        commandValues = new Dictionary<int, string>
        {
            { 1, "wink" },
            { 2, "double blink" },
            { 4, "close your eyes" },
            { 8, "jump" }
        };
    }

    public static string[] Commands(int commandValue) =>    
        commandValue.ReverseOrder()
            ? commandValue.GetCommands().Reverse().ToArray()
            : commandValue.GetCommands().ToArray();

    private static bool ReverseOrder(this int commandValue) =>
        (commandValue & 16) != 0;

    private static IEnumerable<string> GetCommands(this int commandValue)
    {
        foreach (var commands in commandValues)
        {
            if ((commandValue & commands.Key) != 0)
                yield return commands.Value;
        }
    }
}

File: fsharp/secret-handshake/SecretHandshake.fs
module SecretHandshake

let commandsMapping = 
    [
        (1 <<< 0, (fun accumulator -> accumulator @ ["wink"]));
        (1 <<< 1, (fun accumulator -> accumulator @ ["double blink"]));
        (1 <<< 2, (fun accumulator -> accumulator @ ["close your eyes"]));
        (1 <<< 3, (fun accumulator -> accumulator @ ["jump"]))
        (1 <<< 4, (fun accumulator -> accumulator |> List.rev))
    ]

let applyCommand number accumulator (mask, apply) =
    if number &&& mask <> 0 then apply accumulator
    else accumulator

let commands number =
    ([], commandsMapping)
    ||> List.fold (applyCommand number)